### ISSUE TITLE
Non-deterministic exceptions in parallel builds when switching from 1.16.0-RC3 to 1.16.0

### ISSUE THREAD

**Issue by sschuberth** [detekt/detekt#3544]
## Expected Behavior

`./gradlew detekt` should not crash in version 1.16.0 for a project using Kotlin 1.4.31.

## Observed Behavior

`./gradlew detekt` crashes in version 1.16.0, which rather non-deterministic exceptions. At first I thought I would be affected by https://github.com/detekt/detekt/issues/3248 as the exception messages are sometimes similar, but I cannot see any relation to parallel builds (as disabling parallel builds doe not help).

Note that *detekt* version 1.16.0-RC3 actually works!

## Steps to Reproduce

1. Clone https://github.com/sschuberth/stan
2. Checkout the `detekt-1.16.0` branch
3. Run `./gradlew detekt --no-build-cache --rerun-tasks`
4. Observe:

```
sebastian@passau MINGW64 ~/Development/GitHub/sschuberth/stan (detekt-1.16.0)
$ ./gradlew detekt --no-build-cache --rerun-tasks

> Configure project :gui
Project :gui => no module-info.java found

> Task :lib:detekt FAILED
> Task :gui:detekt FAILED
> Task :detekt FAILED

FAILURE: Build completed with 3 failures.

1: Task failed with an exception.
-----------
* What went wrong:
Execution failed for task ':lib:detekt'.
> Collection is empty.

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.
==============================================================================

2: Task failed with an exception.
-----------
* What went wrong:
Execution failed for task ':gui:detekt'.
> kotlin/KotlinNullPointerException

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.
==============================================================================

3: Task failed with an exception.
-----------
* What went wrong:
Execution failed for task ':detekt'.
> io/github/detekt/tooling/api/InvalidConfig

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.
==============================================================================

* Get more help at https://help.gradle.org

Deprecated Gradle features were used in this build, making it incompatible with Gradle 7.0.
Use '--warning-mode all' to show the individual deprecation warnings.
See https://docs.gradle.org/6.8.3/userguide/command_line_interface.html#sec:command_line_warnings

BUILD FAILED in 5s
4 actionable tasks: 4 executed
```

Now do the same on the `detekt-1.16.0-RC3` branch and observe:

```
sebastian@passau MINGW64 ~/Development/GitHub/sschuberth/stan (detekt-1.16.0-RC3)
$ ./gradlew detekt --no-build-cache --rerun-tasks

> Configure project :gui
Project :gui => no module-info.java found

Deprecated Gradle features were used in this build, making it incompatible with Gradle 7.0.
Use '--warning-mode all' to show the individual deprecation warnings.
See https://docs.gradle.org/6.8.3/userguide/command_line_interface.html#sec:command_line_warnings

BUILD SUCCESSFUL in 7s
4 actionable tasks: 4 executed
```

## Context

I'm trying to upgrade from *detekt* 1.15.0 / 1.16.0-RC3 to version 1.16.0 final.

## Your Environment
* Version of detekt used: 1.15.0 / 1.16.0-RC3 and 1.16.0
* Version of Gradle used (if applicable): 6.8.3
* Gradle scan link (add `--scan` option when running the gradle task): n/a
* Operating System and version: Windows 10 64-bit
* Link to your project (if it's a public repository): https://github.com/sschuberth/stan

**Comment by sschuberth** [comment]
My hunch is that https://github.com/detekt/detekt/pull/3521 which revert's *detekt*'s Kotlin version from 1.4.31 to 1.4.21 is the culprit.

**Comment by sschuberth** [comment]
For some reason, it does not even help to use *detekt* 1.16.0 but add

```Kotlin
configurations {
    "detektPlugins" {
        resolutionStrategy {
            force("org.jetbrains.kotlin:kotlin-compiler-embeddable:1.4.31")
            force("org.jetbrains.kotlin:kotlin-stdlib-jdk8:1.4.31")
        }
    }
}
```

to `build.gradle.kts`, even thought that makes the `detektPlugins` configuration's dependencies basically the same as with *detekt* 1.16.0-RC3. Weird.

**Comment by cortinico** [comment]
> After looking at [the changes between versions](https://github.com/detekt/detekt/compare/v1.16.0-RC3...v1.16.0), my hunch is that #3521 which revert's _detekt_'s Kotlin version from 1.4.31 to 1.4.21 is the culprit.

Ideally we could release a 1.16.1 with a Kotlin bump to 1.4.32 (that is not out yet). We're currently blocked on 1.4.31 due to https://youtrack.jetbrains.com/issue/KT-45007

**Comment by sschuberth** [comment]
I'm still wondering why mixing Kotlin 1.4.21 with 1.4.31 is such a problem, it's a patch-level difference only after all. Doesn't Kotlin use semantic versioning?

**Comment by cortinico** [comment]
> I'm still wondering why mixing Kotlin 1.4.21 with 1.4.31 is such a problem, it's a patch-level difference only after all. Doesn't Kotlin use semantic versioning?

They do, but 1.4.31 contained a bug (see https://youtrack.jetbrains.com/issue/KT-45007#focus=Comments-27-4735847.0-0) so we had to revert it.

**Comment by sschuberth** [comment]
I believe we're talking past each other. I just wonder why *detekt* 1.16.0, which uses Kotlin 1.4.21, does not work with my project, which uses Kotlin 1.4.31, as I believe Kotlin 1.4.21 and 1.4.31 should be compatible to each other.

**Comment by cortinico** [comment]
> I just wonder why _detekt_ 1.16.0, which uses Kotlin 1.4.21, does not work with my project, which uses Kotlin 1.4.31

Are we confusing build time dependencies with runtime/compile dependencies?
The fact that your project uses Kotlin 1.4.31 is unrelated to the Kotlin version that Detekt uses while running in your Gradle build (1.4.21).  

> I believe Kotlin 1.4.21 and 1.4.31 should be compatible to each other.

They are definitely compatible. But if you end up in the situation reported by https://github.com/detekt/detekt/pull/3509 and you have Kotlin 1.4.31 in your build dependencies => Then Detekt will fail. That's why we had to revert the version to 1.4.21.

**Comment by sschuberth** [comment]
> Are we confusing build time dependencies with runtime/compile dependencies?

Not sure, maybe. To me build and compile dependencies are the same thing: Dependencies that the app requires at build / compile time (whereas runtime dependencies are those needed for the app to run). But I guess for you "build time dependencies" are dependencies required by the build system, i.e. plugin dependencies.

> The fact that your project uses Kotlin 1.4.31 is unrelated to the Kotlin version that Detekt uses while running in your Gradle build (1.4.21).

So, are you saying the *real* problem is that Gradle 6.8.x uses Kotlin 1.4.20, and that's why *detekt* (as a Gradle plugin) runs into trouble when using Kotlin 1.4.31?

> But if you end up in the situation reported by #3509

Hmm, I don't see any "situation" being reported there. But I guess you meant https://github.com/detekt/detekt/pull/3521 which reverts the change anyway.

So finally, looking at https://github.com/detekt/detekt/issues/3514, it seems something with always be broken for someone. For me, 1.16.0-RC3 works just fine, while 1.16.0 does not.

**Comment by cortinico** [comment]
I've just checked out your project on the `detekt-1.16.0` branch and tried:

```
./gradlew detekt --no-build-cache --rerun-tasks
```

It works fine for me 👍  Here a build scan: https://scans.gradle.com/s/hcapogrquwjva
 
> Not sure, maybe. To me build and compile dependencies are the same thing: Dependencies that the app requires at build / compile time (whereas runtime dependencies are those needed for the app to run). But I guess for you "build time dependencies" are dependencies required by the build system, i.e. plugin dependencies.

To clarify I meant the same distinction that you can find in the Gradle Scan. [Dependencies](https://scans.gradle.com/s/hcapogrquwjva/dependencies) are those used by your project while [Build Dependencies](https://scans.gradle.com/s/hcapogrquwjva/build-dependencies) are those used by your Gradle build and plugins.

> Hmm, I don't see any "situation" being reported there. But I guess you meant #3521 which reverts the change anyway.
> So finally, looking at #3514, it seems something with always be broken for someone. For me, 1.16.0-RC3 works just fine, while 1.16.0 does not.

Yup sorry I linked the wrong issue :) I've edited my link.

> So, are you saying the real problem is that Gradle 6.8.x uses Kotlin 1.4.20, and that's why detekt (as a Gradle plugin) runs into trouble when using Kotlin 1.4.31?

I'm unsure where you problem is. If we can't reproduce is really hard to say at this point. We were able to reproduce the issue in #3514 and we pinpointed the issue in Kotlin 1.4.31 (therefore we reverted the bump before releasing).

I'd say the best to would be if you could update your project to try to reproduce the issue (perhaps using Github Actions?)

**Comment by sschuberth** [comment]
> It works fine for me 👍

That's really unexpected 😮 I guess then one of my `~/.gradle/gradle.properties` settings cause the issue:

```
$ cat ~/.gradle/gradle.properties
org.gradle.caching=true
org.gradle.jvmargs=-Xmx1536M
org.gradle.parallel=true
org.gradle.vfs.watch=true
```

I tried them on by one, and removing `org.gradle.parallel=true` did the trick. So looks like we're back at https://github.com/detekt/detekt/issues/3248 after all...

**Comment by chao2zhang** [comment]
Would you mind sharing the build scan with `./gradlew detekt --rerun-tasks --no-daemon --info -Pdetekt-dry-run=true`. I am highly suspicious of anything wonky on the classpath. E.g. `java.lang.NoClassDefFoundError: Could not initialize class org.jetbrains.kotlin.config.JvmTarget` indicates bytecode mismatch.

**Comment by sschuberth** [comment]
> This is my successful build scan: https://gradle.com/s/2bztsx6vl5se2

Your build has `Parallel=Off`, which succeeds for me, too. Could *you* try with `-Dorg.gradle.parallel=true`?

> Would you mind sharing the build scan with `./gradlew detekt --rerun-tasks --no-daemon --info -Pdetekt-dry-run=true`.

Here it is: https://gradle.com/s/keuhzf2qxhuyk. So, it succeeded, but if I remove `-Pdetekt-dry-run=true` it again fails.

**Comment by chao2zhang** [comment]
I can reproduce the some failure when running with `-Dorg.gradle.parallel=true` https://gradle.com/s/pghkg5ndnfdgu

**Comment by chao2zhang** [comment]
I downgraded the Kotlin version from 1.4.31 to 1.4.21, with `-Dorg.gradle.parallel=true` it seems to pass now https://gradle.com/s/pmtfpcise5maw

**Comment by chao2zhang** [comment]
with `-Dorg.gradle.parallel=true`
- kotlin = 1.4.21 works
- kotlin = 1.4.30 fails `> kotlin/io/CloseableKt` https://gradle.com/s/yuzyfsy7ayj32
- kotlin = 1.4.31 fails `> kotlin/KotlinNullPointerException` https://gradle.com/s/jxpqeo74og3zm

**Comment by sschuberth** [comment]
But as originally written, for me the exceptions thrown are non-deterministic. Running the same build (with Kotlin 1.4.31) again sometimes besides `kotlin/KotlinNullPointerException` also yields `kotlin/UninitializedPropertyAccessException`, `kotlin/jvm/internal/Intrinsics`, `kotlin/jvm/internal/Intrinsics` and others for me.

**Comment by BraisGabin** [comment]
Should we downgrade kotlin version and create a patch version? This way we have more time to fix this.

**Comment by sschuberth** [comment]
You mean *upgrade* the Kotlin version in *detekt* again and do a patch release? While it would be nice to have a "final" version with Kotlin 1.4.31 compatibility, I think it would also be fine to instruct users who need that compatibility to use version 1.16.0-RC3 for now, as the difference to final version 1.16.0 is rather small.

**Comment by cortinico** [comment]
> You mean _upgrade_ the Kotlin version in _detekt_ again and do a patch release?

This. Or ideally wait for `1.4.32` and do a patch release.
As an alternative we can also re-bump to `1.4.31` and invite users that need it to use the latest snapshots till we get a newer stable Kotlin release.

**Comment by sschuberth** [comment]
I'm not even certain that Kotlin 1.4.32 would fix the issue described here, as in none of the stack traces I've seen to far for *this* issue `KeyedExtensionCollector.getPoint()` is present.

**Comment by chao2zhang** [comment]
I guess there are two distinct scenarios:
1. `-Dorg.gradle.parallel=true`
2. `detekt.parallel=true`

It looks like #3248 is specifically for `detekt.parallel=true` but this issue is for `-Dorg.gradle.parallel=true`. Is that correct @sschuberth?

**Comment by sschuberth** [comment]
> I guess there are two distinct scenarios:

That distinction didn't occur to me before, but looking at the initial comment in #3248 which has `detekt.parallel=true` I guess you're right. *However*, it seems to me that several commenters in #3248 (including myself) mistakenly referred to the `org.gradle.parallel=true` case (enabled via the user's `gradle.properties`) instead.

So, I agree that we seem to have two (maybe related) issues then, and we should treat #3248 to be about `detekt.parallel=true` and this issue to be about `org.gradle.parallel=true`.

**Comment by sschuberth** [comment]
FYI, I'm still seeing the same issue with Gradle 7.0, Kotlin 1.5 and detekt 1.17.0-RC2.

**Comment by sschuberth** [comment]
Here's the bottom of the stack trace:

```
2021-05-08T11:39:28.779+0200 [ERROR] [org.gradle.internal.buildevents.BuildExceptionReporter] Caused by: java.lang.reflect.InvocationTargetException
2021-05-08T11:39:28.779+0200 [ERROR] [org.gradle.internal.buildevents.BuildExceptionReporter]   at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
2021-05-08T11:39:28.779+0200 [ERROR] [org.gradle.internal.buildevents.BuildExceptionReporter]   at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
2021-05-08T11:39:28.779+0200 [ERROR] [org.gradle.internal.buildevents.BuildExceptionReporter]   at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
2021-05-08T11:39:28.779+0200 [ERROR] [org.gradle.internal.buildevents.BuildExceptionReporter]   at io.gitlab.arturbosch.detekt.invoke.DefaultCliInvoker.invokeCli(DetektInvoker.kt:56)
2021-05-08T11:39:28.780+0200 [ERROR] [org.gradle.internal.buildevents.BuildExceptionReporter]   ... 122 more
2021-05-08T11:39:28.780+0200 [ERROR] [org.gradle.internal.buildevents.BuildExceptionReporter] Caused by: java.lang.NoClassDefFoundError: kotlin/jvm/internal/Intrinsics
2021-05-08T11:39:28.780+0200 [ERROR] [org.gradle.internal.buildevents.BuildExceptionReporter]   at io.gitlab.arturbosch.detekt.cli.Main.buildRunner(Main.kt)
2021-05-08T11:39:28.780+0200 [ERROR] [org.gradle.internal.buildevents.BuildExceptionReporter]   ... 126 more
2021-05-08T11:39:28.780+0200 [ERROR] [org.gradle.internal.buildevents.BuildExceptionReporter] Caused by: java.lang.ClassNotFoundException: kotlin.jvm.internal.Intrinsics
2021-05-08T11:39:28.780+0200 [ERROR] [org.gradle.internal.buildevents.BuildExceptionReporter]   ... 127 more
```

**Comment by sschuberth** [comment]
I wonder whether putting all of `invokeCli` into a `synchronized` block would help...

**Comment by malachid** [comment]
> FYI, I'm still seeing the same issue with Gradle 7.0, Kotlin 1.5 and detekt 1.17.0-RC2.

Same here with 1.17.1.  It works fine locally, but on Jenkins the second module to run detekt fails with an nondeterministic error.
`org/jetbrains/kotlin/com/intellij/openapi/util/Condition`
`org/jetbrains/kotlin/com/intellij/util/containers/ConcurrentRefHashMap$KeyReference`
`kotlin/io/CloseableKt`

**Comment by cortinico** [comment]
Can you folks try with `1.18.0-RC1` and please report if you're experiencing the same issue or not?

**Comment by sschuberth** [comment]
I just tried with versions 1.18.0-RC1 and 1.18.0-RC2, and both seem to finally fix the issue! 🎉

**Comment by cortinico** [comment]
Closing for now. Feel free to comment if the issue is still happening in a future version.
